modules/writing_observer/writing_observer/awe_nlp.py

```diff
@@ -202,52 +204,104 @@ async def process_texts_parallel(texts, options=None):
-    cache = learning_observer.kvs.KVS()
-
     for writing in writing_data:
         text = writing.get('text', '')
         if len(text) == 0:
             continue
+
+        # Creating text hash
         text_hash = 'NLP_CACHE_' + learning_observer.util.secure_hash(text.encode('utf-8'))
         text_cache_data = await cache[text_hash]
+
         if text_cache_data is None:
             text_cache_data = {}
-        writing.update(text_cache_data)
-        missing_options = set(options if options is not None else []).difference(text_cache_data.keys())
-        needed_options.update(missing_options)
-        if len(missing_options) == 0:
-            results.append(writing)
+
+        if 'features_running' not in text_cache_data:
+            features_running = set()
         else:
-            need_processing.append(writing)
-
-    if len(need_processing) > 0:
-        just_the_text = [w.get("text", "") for w in need_processing]
-        annotated_texts = await processor[mode](just_the_text, list(needed_options))
-
-        for annotated_text, single_doc in zip(annotated_texts, need_processing):
-            if annotated_text != "Error":
-                single_doc.update(annotated_text)
-                text_hash = 'NLP_CACHE_' + learning_observer.util.secure_hash(single_doc['text'].encode('utf-8'))
-                new_cache = {k: v for k, v in single_doc.items() if k != 'student'}
-                await cache.set(text_hash, new_cache)
-        results.extend(need_processing)
+            features_running = set(json.loads(text_cache_data['features_running']))
+
+        if 'features_available' not in text_cache_data:
+            text_cache_data['features_available'] = dict()
+
+        # Check if some options are a subset of features_available
+        if text_cache_data['features_available']:
+            found_options = set_options.intersection(text_cache_data['features_available'].keys())
+            writing.update(text_cache_data['features_available'])
+
+            # If all options were found
+            if found_options == set_options:
+                results.append(writing)
+                continue
+
+        # Check if some options are a subset of features_running
+        # features that are needed but are already running
+        run_successful = False
+        not_found = set_options - found_options
+        features_needed_running = set()  # Features that are needed but are already processing
+        if features_running:
+            features_needed_running = not_found.intersection(features_running)
+        if len(features_needed_running) > 0:
+            while True:
+                new_cache = await cache[text_hash]
+                if new_cache['stop_time'] != "running":
+                    run_successful = True
+                    break
+                if (datetime.datetime.now() - datetime.datetime.strptime(new_cache['start_time'], "%Y-%m-%d %H:%M:%S.%f")).total_seconds() > running_features_wait_time:
```

Comment:
We ought to return some error to the user that features were unable to fetched for one reason or another. For now this can be stubbed in while we continue to work on the communication stream.